csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1385,62 +1409,370 @@ class AllocationReuseModifier : private kir::ExprMutator {
   std::unordered_map<kir::Allocate*, kir::Allocate*> old2new_;
 };
 
-//! Set addresses without any re-use of shared memory. This simply scans
-//! through exprs and each time we find an unaliased Allocate align the current
-//! address and increment it by the given amount.
-class NoReuseSharedMemAllocator : kir::IrVisitor {
+Val* alignExpr(Val* addr, int64_t alignment = 16) {
+  if (alignment == 1) {
+    return addr;
+  }
+  auto n_minus_one = IrBuilder::create<Val>(alignment - 1, DataType::Index);
+  return SimplifyingIrBuilder::bitwiseAndExpr(
+      SimplifyingIrBuilder::addExpr(addr, n_minus_one),
+      SimplifyingIrBuilder::bitwiseNotExpr(n_minus_one));
+}
+
+Val* allocSizeBytes(kir::Allocate* alloc) {
+  const auto buffer_dtype = alloc->buffer()->dtype();
+  const auto dtype_size = dataTypeSize(buffer_dtype);
+  auto size = dtype_size == 1
+      ? alloc->size()
+      : SimplifyingIrBuilder::mulExpr(
+            alloc->size(), IrBuilder::create<Val>(dtype_size, DataType::Index));
+  return size;
+}
+
+//! Allocate differently-sized buffers using a single pass where we push
+//! allocations on a stack then pop them after their last read. This only does
+//! outer sharing: inner sharing is only valid for aliasing.
+//!
+//! Consider the following case, where time proceeds from left to right and
+//! memory is laid out in the upward direction by NoReuseSharedMemAllocator:
+//!
+//!                 +-----+
+//!                 |  C  |
+//!       +-----+   +-----+
+//!       |  B  |
+//!   +---+-----+-----+
+//!   |      A        |
+//!   +---------------+
+//!
+//! In this case the A allocation overlaps both B and C so it cannot be re-used
+//! but we can re-use B's memory. If B and C are compatible, they can be
+//! aliased; however, in this class we assume that all eligible aliases are
+//! already set. We can still reuse memory like below (time points are labelled
+//! along the horizontal axis):
+//!
+//!       +-----+   +-----+
+//!       |  B  |   |  C  |
+//!   +---+-----+-----+---+
+//!   |      A        |
+//!   +---------------+
+//!   a   b     c   d e   f
+//!
+//! Note that we might have incomplete information about allocation sizes prior
+//! to runtime, so our approach only places new allocations on top of the
+//! highest active previous allocation. We implement this with a stack of
+//! allocations to which we push and pop at each time point of interest:
+//!
+//!   a: Allocate A at address 0. Push A
+//!   b: Set T=stack.back()=A. Allocate B at address(T)+size(T). Push B
+//!   c: Pop B
+//!   d: Set T=stack.back()=A. Allocate C at address(T)+size(T). Push C
+//!   e: Don't pop A since it is not at back of stack.
+//!   f: Pop C and A (all inactive allocations at top of stack)
+//!
+//! [Reordering Pushes]
```

Comment:
Could you elaborate more about why this is correct in the comment? It took me a while to figure out why this is correct.

My understanding is, if you have a sequence of `FW`s (first writes) and `LR`s (last reads), like
```
FW A, FW B, LR B, FW C, ...
```
And you want to create a sequence of `alloc`s and `dealloc`s that is compatible with the above sequence of `FW` and `LR`, such as:
```
alloc A, alloc B, dealloc B, alloc C, ...
```
or
```
alloc A, alloc B, alloc C, dealloc B, ...
```

The `alloc-dealloc` sequence is compatible with the `FW-LR` sequence if and only if new alloc does not reuse spaces of active alloc, that is, for all `FW i < LR j`, we must also have `alloc i < dealloc j`, where `<` means earlier than.

The reordering rule satisfy this criterion so it is valid.

Am I understanding correctly?

Reply:
I added a bit more to the comments. The constraint is that all of the pushed allocations have to have their first writes after the first sync following the last reads of all previously popped allocations. Since we clear the holding area at each sync, which is also the only place we pop from the stack, this is met regardless of the order we push the new allocations.